src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_TypeInference.cs

```diff
@@ -0,0 +1,1341 @@
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (6,9): warning CS8602: Possible dereference of a null reference.
+                //         x.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(6, 9),
+                // (9,13): warning CS8600: Cannot convert null to non-nullable reference.
+                //         y = null;
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(9, 13));
+        }
+
+        // PROTOTYPE(NullableReferenceTypes): Deconstruction declaration ignores nullability.
+        [Fact(Skip = "TODO")]
+        public void DeconstructionTypeInference_02()
+        {
+            var source =
+@"class C
+{
+    static (string?, string) F() => (string.Empty, string.Empty);
+    static void G()
+    {
+        (var x, var y) = F();
+        x.ToString();
+        y.ToString();
+        x = null;
+        y = null;
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (7,9): warning CS8602: Possible dereference of a null reference.
+                //         x.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(7, 9),
+                // (10,13): warning CS8600: Cannot convert null to non-nullable reference.
+                //         y = null;
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(10, 13));
+        }
+
+        // PROTOTYPE(NullableReferenceTypes): Deconstruction declaration ignores nullability.
+        [Fact(Skip = "TODO")]
+        public void DeconstructionTypeInference_03()
+        {
+            var source =
+@"class C
+{
+    void Deconstruct(out string? x, out string y)
+    {
+        x = string.Empty;
+        y = string.Empty;
+    }
+    static void M()
+    {
+        (var x, var y) = new C();
+        x.ToString();
+        y.ToString();
+        x = null;
+        y = null;
+    }
+}";
+            var comp = CreateStandardCompilation(
+                source,
+                references: new[] { ValueTupleRef, SystemRuntimeFacadeRef },
+                parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (11,9): warning CS8602: Possible dereference of a null reference.
+                //         x.ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "x").WithLocation(11, 9),
+                // (14,13): warning CS8600: Cannot convert null to non-nullable reference.
+                //         y = null;
+                Diagnostic(ErrorCode.WRN_NullAsNonNullable, "null").WithLocation(14, 13));
+        }
```

Comment:
One more suggestion for deconstruction: using the result.
`string x; string? y; var t = ((x, y) = (...)); // then use t.x and t.y`

Reply:
Added.